### PR TITLE
chore: improve notifications tests

### PR DESCRIPTION
**Pull request by dannykopping**
Blocks #13799

- only use Postgres when necessary
- add some more comments to improve readability

### CODE REVIEW

File: coderd/notifications/notifications_test.go [hunk 1]
```diff
@@ -400,14 +395,14 @@ func TestRetries(t *testing.T) {
 func TestExpiredLeaseIsRequeued(t *testing.T) {
 	t.Parallel()
 
-	// setup
+	// SETUP
 	if !dbtestutil.WillUsePostgres() {
-		t.Skip("This test requires postgres")
+		t.Skip("This test requires postgres; it relies on business-logic implemented in the database")
```
Comment (mtojek) [inline coderd/notifications/notifications_test.go:400]: nit: implemented _only_

File: coderd/notifications/notifications_test.go [hunk 2]
```diff
@@ -151,11 +142,8 @@ func TestSMTPDispatch(t *testing.T) {
 func TestWebhookDispatch(t *testing.T) {
 	t.Parallel()
 
-	// setup
-	if !dbtestutil.WillUsePostgres() {
-		t.Skip("This test requires postgres")
-	}
-	ctx, logger, db := setup(t)
+	// SETUP
+	ctx, logger, db := setupInMemory(t)
```
Comment (mtojek) [inline coderd/notifications/notifications_test.go:146]: The problem with this approach is that we are not testing against the real database. I'm convinced if it is a truly safe way to go. Which parts of the implementation are missing in dbmem?

**Review by mafredri — APPROVED** [review]
What's the pro of no longer using Postgres for these tests? I'm unsure of how may PostgreSQL tests remain, but I'd feel a bit uncomfortable if all our testing was in dbmem.

Otherwise changes look good 👍🏻

File: coderd/notifications/notifications_test.go (hunk 2 above)
Comment (dannykopping) [reply coderd/notifications/notifications_test.go:146]: `dbmem` doesn't actually perform the bulk updates, and along with that comes the retry logic; nor does it handle leases or locking.

There still are some tests which test the real database such as `TestBackpressure`, `TestRetries`, and `TestExpiredLeaseIsRequeued`. Each of those test the full scope of enqueue -> dequeue -> update -> requeue -> update.

In this test specifically, the database is not required to be tested - provided the observed behaviour of having a notification be dispatched to a webhook endpoint is achieved; so it's safe to use a mock database.

**Comment by dannykopping** [comment]
> What's the pro of no longer using Postgres for these tests? I'm unsure of how may PostgreSQL tests remain, but I'd feel a bit uncomfortable if all our testing was in dbmem.
> 
> Otherwise changes look good 👍🏻

The main benefit of not using Postgres really is just speed. See also https://github.com/coder/coder/pull/13863#discussion_r1672079008 for further explanation about the tests still using the database.

**Review by mtojek — APPROVED** [review]

**Review by mtojek — APPROVED** [review]
Good to go 👍

**Pull request merged**